extensions/websocket/02_websocketstream.js

```diff
@@ -0,0 +1,302 @@
+        prefix,
+        context: "Argument 1",
+      });
+      options = webidl.converters.WebSocketStreamOptions(options, {
+        prefix,
+        context: "Argument 1",
+      });
+
+      const wsURL = new URL(url);
+
+      if (wsURL.protocol !== "ws:" && wsURL.protocol !== "wss:") {
+        throw new DOMException(
+          "Only ws & wss schemes are allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      if (wsURL.hash !== "" || wsURL.href.endsWith("#")) {
+        throw new DOMException(
+          "Fragments are not allowed in a WebSocket URL.",
+          "SyntaxError",
+        );
+      }
+
+      this[_url] = wsURL.href;
+
+      core.opSync("op_ws_check_permission", this[_url]);
+
+      if (
+        options.protocols.some((x) =>
+          options.protocols.indexOf(x) !== options.protocols.lastIndexOf(x)
+        )
+      ) {
+        throw new DOMException(
+          "Can't supply multiple times the same protocol.",
+          "SyntaxError",
+        );
+      }
+
+      core.opAsync("op_ws_create", {
+        url: this[_url],
+        protocols: options.protocols.join(", "),
+      }).then((create) => {
+        options.abort?.addEventListener("abort", () => this.close());
+
+        this[_rid] = create.rid;
+        const readable = new ReadableStream({
+          pull: async (controller) => {
+            const { kind, value } = await core.opAsync(
+              "op_ws_next_event",
+              this[_rid],
+            );
+
+            switch (kind) {
+              case "string": {
+                controller.enqueue(value);
+                break;
+              }
+              case "binary": {
+                controller.enqueue(value);
+                break;
+              }
+              case "ping": {
+                core.opAsync("op_ws_send", {
+                  rid: this[_rid],
+                  kind: "pong",
+                });
+                break;
+              }
+              case "close": {
+                this[_closed].resolve(value);
+                tryClose(this[_rid]);
+                break;
+              }
+              case "error": {
+                const err = new Error(value);
+                this[_closed].reject(err);
+                controller.error(err);
+                tryClose(this[_rid]);
+                break;
+              }
+            }
+          },
+          cancel: (reason) => this.close(reason),
+        });
+        const writable = new WritableStream({
+          write: async (chunk) => {
+            if (typeof chunk === "string") {
+              await core.opAsync("op_ws_send", {
+                rid: this[_rid],
+                kind: "text",
+                text: chunk,
+              });
+            } else if (chunk instanceof Uint8Array) {
+              await core.opAsync("op_ws_send", {
+                rid: this[_rid],
+                kind: "binary",
+              }, chunk);
+            }
```

Comment:
```suggestion
            } else {
              // throw TypeError
            }
```
Or is there something that guarantees it's never anything else? (Apart from the `.d.ts` file, I mean.)